lib/libc/picolibc/CMakeLists.txt

```diff
@@ -1,7 +1,17 @@
 # SPDX-License-Identifier: Apache-2.0
 
 zephyr_library()
-zephyr_library_sources(libc-hooks.c)
+zephyr_library_property(PARTIAL TRUE)
```

Comment:
it seems wrong to me that we are starting to modify the `zephyr_library()` this way.

`zephyr_library()` is partly a convenience for users, but also to ensure that such libraries are included inside the `--whole-archive` for various reasons.

Libraries that explicitly should not go inside the `--whole-archive` flags should be created as regular CMake libraries, just like it's done for the `kernel` lib.
https://github.com/zephyrproject-rtos/zephyr/blob/998e839d6743c3ae2633e653415a2ae56bb7fb34/kernel/CMakeLists.txt#L96

and then regular CMake linking should be used.

From the PR I don't see the exact issues faced with circular dependencies, but from the proposed solution then i'm assuming that `kernel` uses / references picolibc ( or newlib) which again uses kernel, thus forming a circular dependency.

Such cases is supported natively in CMake, by having the n-static libs linking to each other, like this:
```
target_link_libraries(kernel PUBLIC picolibc) # Kernel depends on picolibc
target_link_libraries(picolibc PUBLIC kernel) # Picolibc depends on kernel
```
thus informing CMake that those two libs should be linking si that circular dependencies are resolved.

Now, if that isn;t sufficient because one link result in new function references to be resolved, then the `INTERFACE_LINK_LIBRARIES` property can be set in addition.
For example: 
```
set_target_properties(kernel PROPERTIES _LINK_INTERFACE_MULTIPLICITY 3) 
set_target_properties(picolibc PROPERTIES _LINK_INTERFACE_MULTIPLICITY 3) 
```

Reply:
Thanks! I'll give these suggestions a try.

Reply:
This was going fairly well; I've converted libc/common and libc/picolibc to use add_library and reverted all of the kludges related to zephyr_library usage. However, I've run up against a snag -- I need to include the C library in the repeated library set, so that I get:
```
    zephyr/lib/libc/picolibc/libc_picolibc.a \
    zephyr/lib/libc/common/libc_common.a \
    -lc \
    -lgcc \
    zephyr/kernel/libkernel.a \
    zephyr/lib/libc/picolibc/libc_picolibc.a \
    zephyr/lib/libc/common/libc_common.a \
    -lc \
    -lgcc \
    zephyr/kernel/libkernel.a \
    zephyr/lib/libc/picolibc/libc_picolibc.a \
    zephyr/lib/libc/common/libc_common.a \
```
If I could compute the pathname to libc.a, I'd use that directly. However, that's buried inside the picolibc.specs file and gcc won't print it with `-print-file-name` because that doesn't parse the specs file looking for library paths set within that file.

How can I get cmake to insert `-lc -lgcc` multiple times into the link command line? Alternatively, how can I get cmake to wrap those static libraries in `-Wl,--start-group ... -Wl,--end-group`?

Reply:
I also tried using `CMAKE_C_LINK_GROUP_USING_RESCAN`, but cmake gets very upset because it generates a circular dependency between kernel and the header generation bits, the latter not being a static library.

Reply:
Ok, I think this is all generic cmake stuff now. There's a bit of a kludge to make a toolchain C library work correctly, but otherwise it's not terrible. Let me know if this is the direction you're thinking. I'm finding some useful size reductions here.